File: game.cpp
#include "game.h"

namespace oth {

// 8 directions
static const int DR[8] = {-1, -1, -1, 0, 0, 1, 1, 1};
static const int DC[8] = {-1, 0, 1, -1, 1, -1, 0, 1};

void Othello::reset() {
    std::fill(b.begin(), b.end(), (int8_t)EMPTY);
    int c = n / 2;
    b[(c - 1) * n + (c - 1)] = WHITE;
    b[(c - 1) * n + c] = BLACK;
    b[c * n + (c - 1)] = BLACK;
    b[c * n + c] = WHITE;
    player = BLACK;
}

// Does placing `who` at cell idx bracket a line of opponent discs in (dr,dc)?
bool Othello::flips_in_dir(int idx, int who, int dr, int dc) const {
    int r = idx / n, c = idx % n;
    r += dr; c += dc;
    int seen = 0;
    while (r >= 0 && r < n && c >= 0 && c < n) {
        int8_t v = b[r * n + c];
        if (v == -who) { seen++; r += dr; c += dc; }
        else if (v == who) return seen > 0;   // own disc closes the bracket
        else break;                            // empty -> no bracket
    }
    return false;
}

bool Othello::has_move(int who) const {
    for (int idx = 0; idx < n * n; ++idx) {
        if (b[idx] != EMPTY) continue;
        for (int d = 0; d < 8; ++d)
            if (flips_in_dir(idx, who, DR[d], DC[d])) return true;
    }
    return false;
}

bool Othello::is_terminal() const {
    return !has_move(BLACK) && !has_move(WHITE);
}

std::vector<int> Othello::legal_actions() const {
    std::vector<int> out;
    for (int idx = 0; idx < n * n; ++idx) {
        if (b[idx] != EMPTY) continue;
        for (int d = 0; d < 8; ++d)
            if (flips_in_dir(idx, player, DR[d], DC[d])) { out.push_back(idx); break; }
    }
    if (out.empty() && has_move(-player)) out.push_back(PASS());  // must pass
    return out;                                                   // empty => terminal
}

void Othello::apply(int action) {
    if (action == PASS()) { player = -player; return; }
    int who = player;
    b[action] = (int8_t)who;
    int r0 = action / n, c0 = action % n;
    for (int d = 0; d < 8; ++d) {
        if (!flips_in_dir(action, who, DR[d], DC[d])) continue;
        int r = r0 + DR[d], c = c0 + DC[d];
        while (b[r * n + c] == -who) {           // flip the bracketed opponents
            b[r * n + c] = (int8_t)who;
            r += DR[d]; c += DC[d];
        }
    }
    player = -player;
}

int Othello::disc_diff(int who) const {
    int s = 0;
    for (int8_t v : b) s += (v == who) - (v == -who);
    return s;
}

float Othello::terminal_value() const {
    int d = disc_diff(player);
    return d > 0 ? 1.f : (d < 0 ? -1.f : 0.f);
}

void Othello::encode(std::vector<float>& planes) const {
    planes.assign((size_t)3 * n * n, 0.f);
    int hw = n * n;
    for (int i = 0; i < hw; ++i) {
        if (b[i] == player) planes[i] = 1.f;          // plane 0: my stones
        else if (b[i] == -player) planes[hw + i] = 1.f;  // plane 1: opponent
        planes[2 * hw + i] = 1.f;                       // plane 2: all-ones bias
    }
}

}  // namespace oth

File: python/game.py
"""Othello / Reversi rules, NumPy only. Python twin of game.cpp.

Board size is configurable (6 by default for fast CPU self-play, 8 for real
Othello). When encoded for the net the board is always seen from the side to
move, so the policy/value net only ever learns "my stones vs your stones".
"""

import numpy as np

BLACK = +1   # moves first
WHITE = -1
EMPTY = 0

# 8 directions
_DIRS = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]


class Othello:
    def __init__(self, size=6):
        self.n = size
        self.b = np.zeros(size * size, np.int8)   # row-major, values in {EMPTY,BLACK,WHITE}
        self.player = BLACK
        self.reset()

    def A(self):
        return self.n * self.n + 1               # actions: cells + one "pass"

    def PASS(self):
        return self.n * self.n

    def clone(self):
        g = Othello.__new__(Othello)
        g.n, g.b, g.player = self.n, self.b.copy(), self.player
        return g

    def reset(self):
        self.b[:] = EMPTY
        n, c = self.n, self.n // 2
        self.b[(c - 1) * n + (c - 1)] = WHITE
        self.b[(c - 1) * n + c] = BLACK
        self.b[c * n + (c - 1)] = BLACK
        self.b[c * n + c] = WHITE
        self.player = BLACK

    # Does placing `who` at idx bracket a line of opponents in direction (dr,dc)?
    def _flips_in_dir(self, idx, who, dr, dc):
        n = self.n
        r, c = idx // n + dr, idx % n + dc
        seen = 0
        while 0 <= r < n and 0 <= c < n:
            v = self.b[r * n + c]
            if v == -who:
                seen += 1; r += dr; c += dc
            elif v == who:
                return seen > 0                  # own disc closes the bracket
            else:
                break                            # empty -> no bracket
        return False

    def has_move(self, who):
        n = self.n
        for idx in range(n * n):
            if self.b[idx] != EMPTY:
                continue
            for dr, dc in _DIRS:
                if self._flips_in_dir(idx, who, dr, dc):
                    return True
        return False

    def is_terminal(self):
        return not self.has_move(BLACK) and not self.has_move(WHITE)

    def legal_actions(self):
        n = self.n
        out = []
        for idx in range(n * n):
            if self.b[idx] != EMPTY:
                continue
            for dr, dc in _DIRS:
                if self._flips_in_dir(idx, self.player, dr, dc):
                    out.append(idx); break
        if not out and self.has_move(-self.player):
            out.append(self.PASS())              # must pass
        return out                               # empty => terminal

    def apply(self, action):
        if action == self.PASS():
            self.player = -self.player; return
        n, who = self.n, self.player
        self.b[action] = who
        r0, c0 = action // n, action % n
        for dr, dc in _DIRS:
            if not self._flips_in_dir(action, who, dr, dc):
                continue
            r, c = r0 + dr, c0 + dc
            while self.b[r * n + c] == -who:      # flip the bracketed opponents
                self.b[r * n + c] = who
                r += dr; c += dc
        self.player = -self.player

    def disc_diff(self, who):
        return int(np.sum(self.b == who) - np.sum(self.b == -who))

    def terminal_value(self):
        d = self.disc_diff(self.player)          # from side-to-move's perspective
        return 1.0 if d > 0 else (-1.0 if d < 0 else 0.0)

    def encode(self):
        """3 planes (my stones, opponent stones, all-ones bias) as (3,n,n)."""
        n = self.n
        board = self.b.reshape(n, n)
        planes = np.zeros((3, n, n), np.float64)
        planes[0] = (board == self.player)
        planes[1] = (board == -self.player)
        planes[2] = 1.0
        return planes
